src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.cpp

```diff
@@ -158,151 +158,115 @@ Array<State>                    Propagated::calculateStatesAt               (
+        if (instants.isEmpty())
         {
+            continue ;
+        }
 
-            // Call numerical solver to perform propagation
-            Array<State> propagatedNegativeStateVectorArray = propagator_.calculateStatesAt(startState, desiredNegativeInstantArray);
-            sortedStateArray.add(propagatedNegativeStateVectorArray) ;
+        // Forward propagation
+        Array<State> forwardStates = propagator_.calculateStatesAt(this->cachedStateArray_[i], instants) ;
 
-        }
+        // Backward propagation
+        Array<State> backwardStates = propagator_.calculateStatesAt(this->cachedStateArray_[i + 1], instants) ;
+
+        Real durationBetweenStates = (nextStateInstant - thisStateInstant).inSeconds() ;
 
-        // Execute propagation in forwards time
-        if (!desiredPositiveInstantArray.empty())
+        // Take weighted average
+        Array<State> averagedStates = Array<State>::Empty() ;
+        averagedStates.reserve(instants.getSize()) ;
+
+        for (Size k = 0 ; k < instants.getSize() ; ++k)
         {
 
-            // Call numerical solver to perform propagation
-            Array<State> propagatedPositiveStateVectorArray = propagator_.calculateStatesAt(startState, desiredPositiveInstantArray) ;
-            sortedStateArray.add(propagatedPositiveStateVectorArray) ;
+            Real forwardWeight = (instants[k] - this->cachedStateArray_[i].getInstant()).inSeconds() / durationBetweenStates ;
+            Real backwardWeight = (this->cachedStateArray_[i + 1].getInstant() - instants[k]).inSeconds() / durationBetweenStates ;
+
+            VectorXd coordinates = (forwardStates[k].getCoordinates() * forwardWeight + backwardStates[k].getCoordinates() * backwardWeight);
+
+            averagedStates.add({
+                instants[k],
+                Position::Meters({coordinates[0], coordinates[1], coordinates[2]}, gcrfSPtr),
+                Velocity::MetersPerSecond({coordinates[3], coordinates[4], coordinates[5]}, gcrfSPtr)
+            }) ;
 
         }
 
+        allStates.add(averagedStates) ;
+
     }
 
-    // Desort the sorted state array to return it to user as was requested
-    Array<State> unsortedStateArray = Array<State>(anInstantArray.getSize(), State::Undefined()) ;
+    // Propagate any remaining instants forward in time
 
-    for (Size k = 0; k < anInstantArray.getSize(); k++)
+    instants = Array<Instant>::Empty() ;
+
+    for (; j < anInstantArray.getSize() ; ++j)
     {
-        unsortedStateArray[instantArrayPairs[k].second] = sortedStateArray[k] ;
+        instants.add(anInstantArray[j]) ;
     }
 
-    // Add newly propagated states to the cachedStateArray and then resort
-    cachedStateArray_.add(sortedStateArray) ;
-
-    // Re-sort and sanitize cachedStateArray of duplicates and duplicate instants that have different positions/velocities
-    sanitizeCachedArray() ;
+    allStates.add(propagator_.calculateStatesAt(this->cachedStateArray_.accessLast(), instants)) ;
```

Comment:
Yeah we don't update the local cached state array. I don't think we should necessarily update the local cached state array, I think it can eventually become really large. And then we have to get into memory management (based on the use cases). If the user wants to update it, we can provide a setter. That's a better path. Which then also `sanitizes` so we're protected. That way the user can:
- take a cached state array
- propgate to some instants, get a new state array
- join the two and set them if they want
- propagate again
If we find in the future this is a pattern we find ourselves repeating a lot then we can include it in `calculateStatesAt`. But I don't foresee this happening.

Reply:
> If the user wants to update it, we can provide a setter
Okay, I like that! At least we can play with it and use it and that will give us a good appreciation of the point we make here:
> get into memory management (based on the use cases)
But that way we have a possibility (on user side) to do both (update, not update)